app/src/main/java/org/thoughtcrime/securesms/components/mention/MentionRendererDelegate.java

```diff
@@ -0,0 +1,78 @@
+
+import androidx.annotation.ColorInt;
+import androidx.annotation.NonNull;
+import androidx.core.graphics.drawable.DrawableCompat;
+
+import org.thoughtcrime.securesms.R;
+import org.thoughtcrime.securesms.util.ContextUtil;
+import org.thoughtcrime.securesms.util.DrawableUtil;
+import org.thoughtcrime.securesms.util.ViewUtil;
+
+/**
+ * Encapsulates the logic for determining the type of mention rendering needed (single vs multi-line) and then
+ * passing that information to the appropriate {@link MentionRenderer}.
+ * <p></p>
+ * Ported and modified from https://github.com/googlearchive/android-text/tree/master/RoundedBackground-Kotlin
+ */
+public class MentionRendererDelegate {
+
+  private final MentionRenderer single;
+  private final MentionRenderer multi;
+  private final int             horizontalPadding;
+  private final Drawable        drawable;
+  private final Drawable        drawableLeft;
+  private final Drawable        drawableMid;
+  private final Drawable        drawableEnd;
+
+  public MentionRendererDelegate(@NonNull Context context, @ColorInt int tint) {
+    this.horizontalPadding = ViewUtil.dpToPx(2);
+
+    drawable     = DrawableUtil.tint(ContextUtil.requireDrawable(context, R.drawable.mention_text_bg), tint);
+    drawableLeft = DrawableUtil.tint(ContextUtil.requireDrawable(context, R.drawable.mention_text_bg_left), tint);
+    drawableMid  = DrawableUtil.tint(ContextUtil.requireDrawable(context, R.drawable.mention_text_bg_mid), tint);
+    drawableEnd  = DrawableUtil.tint(ContextUtil.requireDrawable(context, R.drawable.mention_text_bg_right), tint);
+
+    single = new MentionRenderer.SingleLineMentionRenderer(horizontalPadding,
+                                                           0,
+                                                           drawable);
+
+    multi = new MentionRenderer.MultiLineMentionRenderer(horizontalPadding,
+                                                         0,
+                                                         drawableLeft,
+                                                         drawableMid,
+                                                         drawableEnd);
+  }
+
+  public void draw(@NonNull Canvas canvas, @NonNull Spanned text, @NonNull Layout layout) {
+    Annotation[] annotations = text.getSpans(0, text.length(), Annotation.class);
+    for (Annotation annotation : annotations) {
+      if (MentionAnnotation.isMentionAnnotation(annotation)) {
+        int spanStart = text.getSpanStart(annotation);
+        int spanEnd   = text.getSpanEnd(annotation);
+        int startLine = layout.getLineForOffset(spanStart);
+        int endLine   = layout.getLineForOffset(spanEnd);
+
+        int startOffset = (int) (layout.getPrimaryHorizontal(spanStart) + -1 * layout.getParagraphDirection(startLine) * horizontalPadding);
+        int endOffset   = (int) (layout.getPrimaryHorizontal(spanEnd) + layout.getParagraphDirection(endLine) * horizontalPadding);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).